### DISCUSSION TITLE
Prepared statements

### DISCUSSION THREAD

**Discussion by dvarrazzo** [psycopg/psycopg#21] [Ideas]
[meta: first time using discussions to check if they gather more participation than ML threads].

The one thing, the most requested thing in psycopg, is support for prepared statements.

In psycopg3 for the moment there is:

- very low level support for prepared statement, i.e. wrapping of libpq functions such as PQsendPrepare/PQsendQueryPrepared (https://www.postgresql.org/docs/current/libpq-async.html#LIBPQ-PQSENDPREPARE)
- automatic use of prepared statements in `cursor.executemany()`, which might eventually stop sucking.

Gathering some ideas:

Prepared statements in the server are per session, so any form of cache is better connected to the connection than the cursor, although the cursors are the obvious interface to give commands.

[In the past](https://gist.github.com/dvarrazzo/3797445) I thought about exposing explicit prepare/deallocate on the cursor, and it was a single prepared query per cursor. A `cursor.prepare(query)` with no args doesn't have types information though: if any it should take an optional array of parameters to get types from.

What I'm thinking about is to prepare queries automatically with a schema such:

- decisions are made after the query is transformed to postgres format (i.e. it is reduced to bytes, all the client-side manipulations have been done, placeholders have been transformed to $ format). There is [an object](https://github.com/psycopg/psycopg3/blob/c790a832/psycopg3/psycopg3/_queries.py#L27) in psycopg3 that takes care of this transformation
- the number of times a query is seen is stored in a LRU cache on the connection
- if a query is seen more than `connection.prepare_threshold` times (proposed default: 5) then it is prepared with the name f'pg3_{hash(query)}' and the following executions are prepared.
- if more than `connection.prepared_number` queries are prepared, the one used least recently is deallocated and evicted from the cache (proposed default: 100).
- Parameters may be fudged on the connection: prepared_threshold=0 would prepare all queries, prepared_threshold=None would disable preparing.
- For the control freak, cursor.execute(query, params, prepare=True) would prepare the query immediately, if it isn't already, prepare=False would avoid preparation. The default None would enable the automatic choice.

What do you think?

**Comment by eddymac** [comment]
Giving very much a personal view.  Most python postgres applications I write are web based and stateless, and often with "forked" processes, so it very much a "Connect->Do->Disconnect->Done" operation, usually with the connection going through pgpool (or similar).  I do have use for prepared statements in that, but are very much cursor based, for instance - processing a batch of invoices: That is preparing statements at beginning of the batch, executing them, once per each invoice, then deallocating the prepared statements at the end.  I am not sure how the automatic preparing descibed would work for me.  Bear in mind, this is a personal view and I am not sure if it is relevant to you.  Thanks.

**Comment by dvarrazzo** [comment]
Prepared statements are now available (they have been for a few days): https://www.psycopg.org/psycopg3/docs/prepared.html

**Reply by DzigaV** [reply]
What's the overhead like on prepared statements? I'm sending the same queries across hundreds of schema, so with this enabled I'd generate rather a lot of prepared statements. I've disabled them for now.

**Comment by dvarrazzo** [comment]
What does sending the same queries across multiple schemas mean? Have you disabled it because you have measured worse performance with prepared queries, and you have measured better performance without?

There is a limit on the prepared statements created, which are managed by a LRU cache, so that no more than 100 queries per connection: running more queries that than would release the least recently used ones.

**Comment by dvarrazzo** [comment]
> no more than 100 queries per connection

More precisely: the connection keeps track of no more than 100 queries, counting how many times they are executed. If some of them are executed more than 5 times then they are prepared; if the connections sees 100 more different queries and the prepared one goes out of the cache then it gets deallocated.

So, at any point during the lifetime of the connection, probably way more than 100 get really prepared, and if the queries executed are always different and more than 100 probably no query ever does.

If you have a querying pattern which preparation makes slower I'm interested to know. As far as I could see, unless you craft a pathological use case (executing a query exactly 6 times and then never anymore, repeating the same pattern hundreds of time), preparing queries only involves a few dict lookups and number increments in Python and I doubt is measurable.

**Reply by DzigaV** [reply]
I'm new to Postgres, forgive me.

So it's safe to assume there is no client overhead. What about on the server? This is the wrong place to ask, I'll do my own research. I wondering how far I can increase this limit, that's all. I vary the schema.table a lot across my queries, so these will all be treated as separate statements, I understand. Now let's say I have 50 long-lived database connections, each preparing their own statements, I imagine there are implications.

**Comment by dvarrazzo** [comment]
Psycopg3 keeps a limit on the number of statements that get prepared, and a few dozens I'm sure is a very conservative limit :slightly_smiling_face:  I am pretty sure you can prepare several hundreds of statements per connection and if your server is well dimensioned (which should if you have 50 connections open) shouldn't even notice.

I am very happy if you do your own research about the topic. Probably the best place to enquire is the pgsql-general mailing list.

In order to see what's going behind the scenes you can look at the `connection._prepared` object and to the table `pg_prepared_statements` on the connection (you won't see the same prepared statements from another connection, obviously).

**Comment by DzigaV** [comment]
I'm loving prepared statements with no effort. The default settings seem to make sense for most users, and I notice they mirror conventions used in other non-Py drivers.

The fine-grained control is very useful for me. I'm immediately preparing all queries that target my public schema, because they're constantly reused. But I can choose not to prepare the rest, because I have hundreds of different SQL statements targeting hundreds of schema, leading to an astronomical number of query variants in constant use.

**Reply by dvarrazzo** [reply]
Thank you for the feedback!

It would be interesting to know if, with the default settings, your pattern of use leads to an sane behaviour or to useless prepare/unprepare.

It could be useful to add usage stats to the prepare object maybe.

**Comment by wieck** [comment]
Coming here because I am interested in installing 3.1dev0 inside a venv to specifically test this feature (prepared statements).

That said, this change in behavior should be prominently highlighted in the release notes as it has the potential to break existing installations in a not so obvious way. Many users deploy external connection pools like pgbouncer, which allows to connect multiple applications through the same connection pool. When pgbouncer (as one example) is configured in 'transaction' mode, it is allowed to serve many clients through fewer database connection and change the client/dbconnection association on transaction boundaries (i.e. after commit). This is of course in conflict with server side prepared statements as not all clients may have the same statements prepared at the same time or under the same name. What makes this especially dangerous is that in my experience (as a PostgreSQL contributor for 25 years) most users don't test their existing applications against new version in a way that would expose this.

As for performance, server side prepared statements are said to have little performance impact on simple, single table statements. However I recently ran a TPC-C style workload (Benchmarksql) configured as saturation load against a VM, that was intentionally configured to become CPU bound. Benchmarksql is implemented in Java. Setting prepareThreshold=0 in the URI caused the throughput (tpmC) to drop by a whooping 31%. It should be noted that the TPC-C workload, although containing about 30 different queries, only has 3 statements that have joins or subselects. Nonetheless, most of them are slightly more complex than a simple, single table select with one qualification. So there is quite some value in this feature.

**Reply by dvarrazzo** [reply]
The incompatibility of prepared statements with pgbouncer [is documented](https://www.psycopg.org/psycopg3/docs/advanced/prepare.html#prepared-statements). Improvements to the documentation are welcome.

[I have tried](https://github.com/pgbouncer/pgbouncer/issues/653) to obtain that pgbouncer advertises its presence but the developer are not interested. I would be very interested in disabling prepared statement automatically, if detected, but I don't think we have the tools for it.

**Reply by wieck** [reply]
@dvarrazzo thanks for the quick reply.

That is a problem that I have observed building in the PostgreSQL community for many years. Not being willing to introduce any features that cater to only a few special cases. Back when I was improving Slony-I the Londiste project joined us and we got the session_replication_role into core. I don't think that would be as easy today.

**Comment by wieck** [comment]
@dvarrazzo May I ask what the reason behind doing a DEALLOCATE ALL on conn.rollback() is?

PostgreSQL (at least the V13 I am using to test here) remembers a successfully prepared statement just fine, even if the transaction subsequently fails or is rolled back for other reasons. Some applications trigger frequent exceptions like foreign key violations as part of their business logic. Clearing out ALL prepared statements and starting from scratch every time that happens doesn't seem necessary.

**Reply by dvarrazzo** [reply]
For the problem [highlighted here](https://github.com/sqlalchemy/sqlalchemy/issues/6842#issuecomment-925131836) and [discussed here](https://www.postgresql.org/message-id/2220999.1632335261%40sss.pgh.pa.us) with the postgres hacker.

Yes, Postgres remembers the prepared statement just fine... actually pretty well. If the objects disappears, it will keep referring to its oid...

Ways an object can disappear are either DROP or ROLLBACK, so we flush the prepared statement cache when we identify that these commands were executed.

**Reply by wieck** [reply]
OK, that makes sense. However, in that case you'd better check the completion code of COMMIT. Because a transaction in aborted state still accepts COMMIT, doesn't raise a new error but instead just reports ROLLBACK as completion code. This is a gripe that the JDBC project has.

**Reply by wieck** [reply]
I tested the SQLAlchemy case with the following against PG 13.5 using today's (2022-02-02 aka Groundhog Day) master branch of psycopg.git:

```
#!/usr/bin/env python3

# Test program to analyze on the wire protocol behavior of
# psycopg 3.1

import sys
import psycopg

def main():
    if len(sys.argv) < 2:
        print("usage: test1.py CONNINFO", file = sys.stderr)
        sys.exit(2)

    conninfo = sys.argv[1]
    conn = psycopg.connect(conninfo)

    for i in range(3):
        conn.execute("""CREATE TYPE an_enum AS ENUM ('foo', 'bar', 'baz');""")
        conn.execute("CREATE TABLE foo(id integer, bar an_enum[])")
        conn.cursor().executemany(
            "INSERT INTO foo (bar) VALUES (%(enum_col)s::an_enum[])",
            [{"enum_col": ["foo"]}, {"enum_col": ["foo", "bar"]}],
        )
        print(conn.execute("select * from foo").fetchall())
        conn.rollback()

if __name__ == '__main__':
    main()
```
    
Which is more or less verbatim their test case. The output was

```
$ ./test3.py "host=pghost user=postgres"
[(None, '{foo}'), (None, '{foo,bar}')]
[(None, '{foo}'), (None, '{foo,bar}')]
[(None, '{foo}'), (None, '{foo,bar}')]
$
```

No error. Something doesn't line up here.

In my manual tests using psql and the PREPARE statement, PG 13.5 would error out with "relation X does not exist" when a prepared statement is executed after a DROP TABLE, but it would replan the statement properly if the object is subsequently recreated. 

My best guess is that the SQLAlchemy test case was run against a PG version that doesn't do dependency tracking and cached plan invalidation as thoroughly, as more recent versions do.

I think more testing is in order before finalizing this behavior. I will run this test against previous versions of PG to see if/when this behavior may have changed.

**Reply by wieck** [reply]
Last but not least, I tested the above against PostgreSQL versions 10.19, 11.14, 12.9, 13.5 and 14.1. All behave like that. No error and just what is expected.

Note that 10 is the oldest currently supported major PostgreSQL version because 9.6 went EOL in November 2021. But I went one further and tested against 9.6.24 with the exact same result. I can't easily test against older versions than that because my ansible driven environment relies on the PGDG repositories, which no longer provide 9.5 and older on CentOS or Rocky.
